### DISCUSSION TITLE
Tests parsing via jest-parse-cli

### DISCUSSION THREAD

**Discussion by marilari88** [nvim-neotest/neotest#113] [Ideas]
Hi @rcarriga, I'm working on improving [neotest-vitest](https://github.com/marilari88/neotest-vitest) parsing logic in order to solve the dynamic tests issue (test.each, template string, etc.). And after reading your reply at #68 I was attempting to parse tests throught to the [jest-parser-cli](https://www.npmjs.com/package/jest-parser-cli) package I published on NPM  (bypassing treesitter). I managed to get parsed JSON object from the library, but I'm struggling to transform the JSON object to the neotest Tree. Maybe there Neotest already provides a way to transform JSON to Tree without using treesitter. Isn't it?
Anyway, what do you think about this solution? 
I noticed two CONS: 
- Parsing is slower than treesitter
- Users have to install `jest-parser-cli` globally
But we can let the user choose in the config file how to parse (treesitter or jest-parser-cli)

**Comment by rcarriga** [comment]
Sounds good! Unfortunately I don't think many runners will be faster than treesitter (especially interpreted languages) just because it's so simple and optimised. I do like the idea of allowing users to choose which parser to use, there's no point suffering the perfomance hit if they're not using dynamic tests.

For converting JSON to a `Tree`, the `Tree.from_list` function is the only intended way to the create a tree so you just need to transform the JSON to the structure required, which is a nested list structure. You can check out the tests in https://github.com/nvim-neotest/neotest/blob/master/tests/unit/lib/positions/init_spec.lua to the see the shape. `Tree:to_list` will produce the required shape to if you want to play around locally with an existing tree.

For getting the package installed for users, what about either packaging the cli with the plugin or having a `package.json` and getting them to `npm install` using their plugin manager when installing (similar to how coc.nvim works)? You'll then just need to do some path finding which isn't too difficult. neotest-python also requires using a python script that it just bundles along with the plugin and does this.

**Comment by jfpedroza** [comment]
We solved this in Elixir by changing how the IDs are generated. We use `relative_path:line` as the ID to make it look like it's a single test.

On the Elixir runner side, we append the output of dynamic tests in a single file and the status is "failed" if one of them failed.

That way you can have dynamic tests working even when using treesitter.

**Reply by rcarriga** [reply]
The only issue with using line numbers in IDs is that they will be invalidated if a line is added/removed above them right? Or am I misunderstanding?

**Reply by jfpedroza** [reply]
Yes, that is correct. Adding a new line before the test makes the sign disappear.
